// @ianvs/prettier-plugin-sort-imports@3.3.1/lib/src/utils/get-code-from-ast.js
"use strict";
var __spreadArray = (this && this.__spreadArray) || function (to, from) {
    for (var i = 0, il = from.length, j = to.length; i < il; i++, j++)
        to[j] = from[i];
    return to;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getCodeFromAst = void 0;
var generator_1 = __importDefault(require("@babel/generator"));
var types_1 = require("@babel/types");
var constants_1 = require("../constants");
var get_all_comments_from_nodes_1 = require("./get-all-comments-from-nodes");
var remove_nodes_from_original_code_1 = require("./remove-nodes-from-original-code");
/**
 * This function generate a code string from the passed nodes.
 * @param nodes All imports, in the sorted order in which they should appear in
 * the generated code.
 * @param originalCode The original input code that was passed to this plugin.
 * @param directives All directive prologues from the original code (e.g.
 * `"use strict";`).
 * @param interpreter Optional interpreter directives, if present (e.g.
 * `#!/bin/node`).
 */
var getCodeFromAst = function (nodes, originalCode, directives, interpreter) {
    var allCommentsFromImports = get_all_comments_from_nodes_1.getAllCommentsFromNodes(nodes);
    var allCommentsFromDirectives = get_all_comments_from_nodes_1.getAllCommentsFromNodes(directives);
    var nodesToRemoveFromCode = __spreadArray(__spreadArray(__spreadArray(__spreadArray(__spreadArray([], nodes), allCommentsFromImports), allCommentsFromDirectives), (interpreter ? [interpreter] : [])), directives);
    var codeWithoutImportsAndInterpreter = remove_nodes_from_original_code_1.removeNodesFromOriginalCode(originalCode, nodesToRemoveFromCode);
    var newAST = types_1.file({
        type: 'Program',
        body: nodes,
        directives: directives,
        sourceType: 'module',
        interpreter: interpreter,
        sourceFile: '',
        leadingComments: [],
        innerComments: [],
        trailingComments: [],
        start: 0,
        end: 0,
        loc: {
            start: { line: 0, column: 0 },
            end: { line: 0, column: 0 },
        },
    });
    var code = generator_1.default(newAST).code;
    return (code.replace(/"PRETTIER_PLUGIN_SORT_IMPORTS_NEW_LINE";/gi, constants_1.newLineCharacters) + codeWithoutImportsAndInterpreter.trim());
};
exports.getCodeFromAst = getCodeFromAst;
